src/crimson/tools/store_nbd/fs_driver.cc

```diff
@@ -0,0 +1,167 @@
+      0,
+      ghobject_t::NO_GEN),
+    offset % config.object_size
+  };
+}
+
+seastar::future<> FSDriver::write(
+  off_t offset,
+  bufferptr ptr)
+{
+  auto mapping = map_offset(offset);
+  ceph_assert(mapping.offset + ptr.length() <= config.object_size);
+  ceph::os::Transaction t;
+  bufferlist bl;
+  bl.append(ptr);
+  t.write(
+    mapping.chandle->get_cid(), 
+    mapping.object,
+    mapping.offset,
+    ptr.length(),
+    bl,
+    0);
+  return fs->do_transaction(
+    mapping.chandle,
+    std::move(t));
+}
+
+seastar::future<bufferlist> FSDriver::read(
+  off_t offset,
+  size_t size)
+{
+  auto mapping = map_offset(offset);
+  ceph_assert((mapping.offset + size) <= config.object_size);
+  return fs->read(
+    mapping.chandle,
+    mapping.object,
+    mapping.offset,
+    size,
+    0
+  ).handle_error(
+    crimson::ct_error::enoent::handle([size](auto &e) {
+      bufferlist bl;
+      bl.append_zero(size);
+      return seastar::make_ready_future<bufferlist>(std::move(bl));
+    }),
+    crimson::ct_error::assert_all{"Unrecoverable error in FSDriver::read"}
+  ).then([this, size](auto &&bl) {
+    if (bl.length() < size) {
+      bl.append_zero(size - bl.length());
+    }
+    return seastar::make_ready_future<bufferlist>(std::move(bl));
+  });
+}
+
+seastar::future<> FSDriver::mkfs()
+{
+  init();
+  assert(fs);
+  return fs->start(
+  ).then([this] {
+    uuid_d uuid;
+    uuid.generate_random();
+    return fs->mkfs(uuid);
+  }).then([this] {
+    return fs->stop();
+  }).then([this] {
+    init();
+    return fs->start();
+  }).then([this] {
+    return fs->mount();
+  }).then([this] {
+    return seastar::do_for_each(
+      boost::counting_iterator<unsigned>(0),
+      boost::counting_iterator<unsigned>(config.num_collections),
+      [this](auto i) {
+	return fs->create_new_collection(get_coll(i)
+	).then([this, i](auto coll) {
+	  ceph::os::Transaction t;
+	  t.create_collection(get_coll(i), 0);
+	  return fs->do_transaction(coll, std::move(t));
+	});
+      });
+  }).then([this] {
+    return fs->umount();
+  }).then([this] {
+    return fs->stop();
+  }).then([this] {
+    fs.reset();
+    return seastar::now();
+  });
+}
+
+seastar::future<> FSDriver::mount()
+{
+  ceph_assert(config.path);
+  return (
+    config.mkfs ? mkfs() : seastar::now()
+  ).then([this] {
+    init();
+    return fs->start();
+  }).then([this] {
+    return fs->mount();
+  }).then([this] {
+    return seastar::do_for_each(
+      boost::counting_iterator<unsigned>(0),
+      boost::counting_iterator<unsigned>(config.num_collections),
+      [this](auto i) {
+	return fs->open_collection(get_coll(i)
+	).then([this, i](auto ref) {
+	  collections[i] = ref;
+	  return seastar::now();
+	});
+      });
+  });
+};
+
+seastar::future<> FSDriver::close()
+{
+  collections.clear();
+  return fs->stop(
+  ).then([this] {
+    return fs->umount();
+  }).then([this] {
+    return fs->stop();
+  }).then([this] {
+    fs.reset();
+    return fs->stat();
```

Comment:
after `fs.reset()`, i think the `FuturizedStore` instance managed by `fs` is a nullptr. so the behavior of `fs->stat()` would be undefined.

Reply:
Yeah, that stat() is supposed to be in mount().  Will fix.